cartridge/roles/coordinator.lua

```diff
@@ -123,119 +108,136 @@ local function control_loop(conn)
-    until not pcall(fiber.testcancel)
-end
-
-local function take_control(uri)
-    checks('string')
-    local conn, err = NetboxConnectError:pcall(netbox.connect, uri)
-    if conn == nil then
-        return nil, err
-    elseif not conn:is_connected() then
-        return nil, NetboxConnectError:new('"%s:%s": %s',
-            conn.host, conn.port, conn.error
-        )
+        fiber.testcancel()
     end
+end
 
-    local lock_delay, err = errors.netbox_call(conn, 'get_lock_delay',
-        nil, {timeout = vars.options.NETBOX_CALL_TIMEOUT}
-    )
-    if lock_delay == nil then
-        return nil, err
-    end
+local function take_control(client)
+    checks('stateboard_client')
 
     local lock_args = {
         confapplier.get_instance_uuid(),
         confapplier.get_advertise_uri()
     }
 
-    local ok, err = errors.netbox_call(conn, 'acquire_lock',
-        lock_args, {timeout = vars.options.NETBOX_CALL_TIMEOUT}
-    )
+    local session = client:get_session()
+    assert(not session:is_locked())
+    assert(session.ctx == nil)
 
-    if ok == nil then
+    local lock_delay, err = session:get_lock_delay()
+    if lock_delay == nil then
         return nil, err
     end
 
-    if ok ~= true then
-        return false
+    while true do
+        local ok, err = session:acquire_lock(lock_args)
+        if ok == nil then
+            return nil, err
+        end
+
+        if not ok then
+            fiber.sleep(lock_delay/2)
+        else
+            break
+        end
     end
 
+    local leaders, err = session:get_leaders()
+    if leaders == nil then
+        return nil, err
+    end
+
+    local ctx = {
+        members = nil,
+        decisions = {}
+    }
+    for replicaset_uuid, leader_uuid in pairs(leaders) do
+        ctx.decisions[replicaset_uuid] = pack_decision(leader_uuid)
+    end
+    session.ctx = ctx
+
     log.info('Lock acquired')
-    vars.conn = conn
-    local control_fiber = fiber.new(control_loop, conn)
+    --------------------------------------------------------------------
+    local control_fiber = fiber.new(control_loop, session)
     control_fiber:name('failover-coordinate')
 
+    -- Warning: fragile code.
+    -- The goal: perform garbage collection when the fiber is cancelled
     repeat
-        -- Warning: fragile code.
-        -- Cancelled fibers raise error on every yield
         if not pcall(fiber.sleep, lock_delay/2) then
             break
         end
 
-        if pcall(fiber.status, control_fiber) == 'dead'
-        or not errors.netbox_call(conn, 'acquire_lock',
-            lock_args, {timeout = vars.options.NETBOX_CALL_TIMEOUT}
-        ) then
+        if fiber.status(control_fiber) == 'dead'
+        or not session:acquire_lock(lock_args)
+        then
             break
         end
     until not pcall(fiber.testcancel)
 
-    pcall(conn.close, conn)
-    pcall(fiber.cancel, control_fiber)
-    vars.conn = nil
+    session:drop()
+    pcall(function()
```

Comment:
pcall(fiber.cancel, control_fiber) looks more appropriate IMO. Feel free to ignore.